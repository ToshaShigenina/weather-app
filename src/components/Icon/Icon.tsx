import React from 'react';
import cn from 'classnames';

import './css/icon.css';

export type IconSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
export type IconType = 'point' | 'search' | 'switch' | 'arrow';
export interface IconProps extends React.SVGAttributes<SVGElement> {
  size?: IconSize
  type: IconType
}

const sizes: Record<IconSize, number> = {
  xl: 64,
  lg: 32,
  md: 24,
  sm: 16,
  xs: 8,
};

const Icon: React.FC<IconProps> = ({
  type,
  size = 'md',
  className,
  width,
  height,
  ...otherProps
}) => {
  const iconWidth = width || sizes[size];
  const iconHeight = height || sizes[size];

  return (
    <svg
      className={ cn(className, 'svg-icon') }
      width={ iconWidth }
      height={ iconHeight }
      { ...otherProps }
    >
      <use xlinkHref={ `#${type}` } />
    </svg>
  )
};

export default Icon;