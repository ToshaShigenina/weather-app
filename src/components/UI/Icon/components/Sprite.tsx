import React, { FC, SVGAttributes, useEffect } from 'react';
import { createPortal } from 'react-dom';
import { ReactComponent as SpriteSvg } from '../svg/sprite.svg'

import '../css/sprite.css';

const portalNode = document.createElement('div');
portalNode.setAttribute('id', 'sprite');
portalNode.setAttribute('data-testid', 'sprite');
portalNode.hidden = true;

export const Sprite: FC<SVGAttributes<SVGElement>> = (props) => {
	useEffect(() => {
		document.body.appendChild(portalNode);
		return () => portalNode.remove();
	});
	return createPortal(<SpriteSvg { ...props } />, portalNode)
};
