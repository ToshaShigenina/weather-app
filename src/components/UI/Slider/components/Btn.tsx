import React, { ButtonHTMLAttributes, FC } from 'react';
import cn from 'classnames';
import { ReactComponent as SliderArrow } from '../svg/slider-arrow.svg'

type SliderBtnDirectionType = 'prev' | 'next';
interface SliderBtnProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    direction: SliderBtnDirectionType
}

export const Btn: FC<SliderBtnProps> = ({ direction = '', className, ...otherProps }) => {
    return (
        <button
            className={ cn(className, 'slider__btn', direction && `slider__btn-${direction}`) }
            { ...otherProps }
        >
            <SliderArrow />
        </button>
    );
};
