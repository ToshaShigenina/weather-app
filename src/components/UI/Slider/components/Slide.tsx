import React, { FC, HTMLAttributes, useContext, useEffect, useState } from 'react';
import cn from 'classnames';
import { SliderContext, type SliderContextType } from '../context';

export const Slide: FC<HTMLAttributes<HTMLDivElement>> = ({ children, className = '', style = {}, ...otherProps }) => {
    const slideKey = Symbol();
    const [slideIndex, setSlideIndex] = useState(0);
    const sliderCtx = useContext<SliderContextType | null>(SliderContext);
    const width = sliderCtx && sliderCtx.width > 0
        ? sliderCtx.width + 'px'
        : '100%';
    const marginLeft = sliderCtx?.margin || 0;

    useEffect(() => {
        if (sliderCtx && sliderCtx.addSliderKey && !sliderCtx.sliderKeys.includes(slideKey)) {
            setSlideIndex(sliderCtx.sliderKeys.length);
            sliderCtx.addSliderKey(slideKey);
        }
    }, []);
    return (
        <div
            className={ cn(className, 'slide') }
            style={ {
                ...style,
                width,
                marginLeft: slideIndex ? marginLeft : 0,
            } }
            { ...otherProps }
        >
            { children }
        </div>
    );
};
