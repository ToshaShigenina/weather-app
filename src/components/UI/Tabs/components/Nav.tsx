import React, { FC, OlHTMLAttributes } from "react";
import cn from "classnames";

export interface NavProps extends Omit<OlHTMLAttributes<HTMLUListElement>, 'children' | 'onClick'> {
	active?: number,
	items?: string[],
	onClick?: (index: number) => void
}

export const Nav: FC<NavProps> = ({ active = 0, items = [], onClick, ...othrProps }) => {
	const onClickHandler = (index: number) => {
		if (onClick) {
			onClick(index);
		}
	};

	return (
		<>
			{
				Boolean(items.length) &&
				<ul className={ 'tabs-nav' } { ...othrProps }>
					{
						items.map((item, i) => {
							return (
								<li className={ cn('tabs-nav__item', active === i && 'active') } key={ 'tabs-nav-item-' + i }>
									<button
										className="tabs-nav__btn"
										onClick={ () => onClickHandler(i) }
									>
										{ item }
									</button>
								</li>
							)
						})
					}
				</ul>
			}
		</>
	);
};
