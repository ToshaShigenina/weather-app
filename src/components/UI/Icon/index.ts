import { Icon as InitialIcon } from './components/Icon';
import { Sprite } from './components/Sprite';

type InitialIconType = typeof InitialIcon;
type CompaundComponentType = InitialIconType & {
	Sprite: typeof Sprite
};

const Icon = InitialIcon as CompaundComponentType;
Icon.Sprite = Sprite;

export { Icon };
export type { IconProps, IconSize, IconType } from './components/Icon';
