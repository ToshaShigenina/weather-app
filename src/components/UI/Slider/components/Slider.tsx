import React, { FC, HTMLAttributes, useRef, useState, useEffect } from 'react';
import { Btn } from './Btn';
import { useResize } from '../../../../hooks';
import { getCountSlides, getMinBreackpoint, getBreackpointsMap } from '../utils'
import { SliderContextProvider, type SliderContextType } from '../context';
import '../style.css';

export type SliderBreackpointProps = {
    slidesPerView?: number,
    marginBetweenSlides?: number,
}
export interface SliderProps extends HTMLAttributes<HTMLDivElement>, SliderBreackpointProps {
    breackpoints?: Record<number, SliderBreackpointProps>,
}

export const Slider: FC<SliderProps> = ({ children = null, slidesPerView = 1, marginBetweenSlides = 0, breackpoints, ...otherProps }) => {
    const sliderWrapperRef = useRef(null);
    const { width: windowWidth } = useResize();
    const [currentSlide, setCurrentSlide] = useState(1);
    const [slideWidth, setSlideWidth] = useState(0);
    const [countSlides, setCountSlides] = useState(0);
    const [countSlidesPerView, setCountSlidesPerView] = useState(slidesPerView);
    const [sliderGap, setSliderGap] = useState(marginBetweenSlides);
    const [translate, setTranslate] = useState(0);
    const [sliderKeys, setSliderKeys] = useState<symbol[]>([]);
    const context: SliderContextType = {
        width: slideWidth,
        margin: sliderGap,
        sliderKeys,
        addSliderKey: (key: symbol) => {
            sliderKeys.push(key);
            setSliderKeys(sliderKeys);
        }
    };

    const calcSlideWidth = () => {
        const count = getCountSlides(children);
        setCountSlides(count);
        if (count > 0 && sliderWrapperRef && sliderWrapperRef.current) {
            const sliderWidth = (sliderWrapperRef.current as HTMLDivElement).clientWidth
            setSlideWidth((sliderWidth - ((countSlidesPerView - 1) * sliderGap)) / countSlidesPerView);
        }
    };
    useEffect(() => {
        if (breackpoints) {
            const bpMap = getBreackpointsMap(breackpoints, windowWidth);
            if (Object.values(bpMap).every((item) => !item)) {
                setCountSlidesPerView(slidesPerView);
                setSliderGap(marginBetweenSlides);
            } else {
                const bpMin = getMinBreackpoint(bpMap);
                const bpTemp = breackpoints[bpMin];
                if (bpTemp.slidesPerView !== undefined) setCountSlidesPerView(bpTemp.slidesPerView);
                if (bpTemp.marginBetweenSlides !== undefined) setSliderGap(bpTemp.marginBetweenSlides);
            }
        }
        calcSlideWidth();
    }, [windowWidth]);
    useEffect(() => {
        calcSlideWidth();
    }, [children, countSlidesPerView, sliderGap]);

    const clickToPrevBtn = () => {
        if (currentSlide > 1) {
            setCurrentSlide(currentSlide - 1)
            setTranslate(translate + (slideWidth + sliderGap));
        }
    };
    const clickToNextBtn = () => {
        if (currentSlide <= countSlides - countSlidesPerView) {
            setCurrentSlide(currentSlide + 1)
            setTranslate(translate - (slideWidth + sliderGap));
        }
    };

    return (
        <SliderContextProvider value={ context }>
            <div className="slider" { ...otherProps }>
                <div className="slider__wrapper" ref={ sliderWrapperRef }>
                    <div
                        className="slider__container"
                        style={ { transform: `translateX(${translate}px)` } }
                    >
                        { children }
                    </div>
                </div>
                <Btn
                    onClick={ clickToPrevBtn }
                    direction="prev"
                    disabled={ currentSlide <= 1 }
                />
                <Btn
                    onClick={ clickToNextBtn }
                    direction="next"
                    disabled={ currentSlide > countSlides - countSlidesPerView }
                />
            </div>
        </SliderContextProvider>
    );
};
