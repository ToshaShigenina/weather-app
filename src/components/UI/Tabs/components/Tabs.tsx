import React, { FC, HTMLAttributes, ReactNode, useState } from "react";
import cn from "classnames";
import { Nav } from './Nav';
import '../style.css';

export type TabItem = {
	label: string
	children: ReactNode
};
export interface TabsProps extends Omit<HTMLAttributes<HTMLDivElement>, 'children'> {
	items: TabItem[]
};
type LabelsFunc = (items: TabItem[]) => string[];

const getLabels: LabelsFunc = (items) => {
	return items.map(item => item.label);
};
export const Tabs: FC<TabsProps> = ({ items = [] }) => {
	const [active, setActive] = useState(0);

	const changeActiveTab = (index: number) => {
		setActive(index);
	};

	return (
		<>
			<Nav active={ active } items={ getLabels(items) } onClick={ changeActiveTab } />
			<div className="tabs-content">
				{ items[active].children }
			</div>
		</>
	);
};
