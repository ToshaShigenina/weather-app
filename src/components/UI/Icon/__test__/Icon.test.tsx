import React from 'react';
import { render, screen } from '@testing-library/react';
import { mountTest } from '../../../../tests';
import { Icon, type IconProps } from '../components/Icon';

const testId = 'icon';

describe('Icon', () => {
	mountTest<IconProps>(Icon, { type: 'point' } as IconProps);

	it('установлены дефолтная ширина и высота', () => {
		render(<Icon type="point" data-testid={ testId } />);
		expect(screen.getByTestId(testId).getAttribute('width')).toEqual('24');
		expect(screen.getByTestId(testId).getAttribute('height')).toEqual('24');
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});
	it('ширина и высота заданы через проп size', () => {
		render(<Icon type="point" size="lg" data-testid={ testId } />);
		expect(screen.getByTestId(testId).getAttribute('width')).toEqual('32');
		expect(screen.getByTestId(testId).getAttribute('height')).toEqual('32');
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});
	it('меняется тип иконки', () => {
		render(<Icon type="switch" data-testid={ testId } />);
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});
	it('добавляется дополнительный класс', () => {
		render(<Icon type="point" className="test-class" data-testid={ testId } />);
		expect(screen.getByTestId(testId)).toHaveClass('svg-icon');
		expect(screen.getByTestId(testId)).toHaveClass('test-class');
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});
	it('задается кастомная ширина и высота', () => {
		render(<Icon type="point" width="100" height="150" data-testid={ testId } />);
		expect(screen.getByTestId(testId).getAttribute('width')).toEqual('100');
		expect(screen.getByTestId(testId).getAttribute('height')).toEqual('150');
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});
});