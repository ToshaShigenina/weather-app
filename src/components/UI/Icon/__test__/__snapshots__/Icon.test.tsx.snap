// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon добавляется дополнительный класс 1`] = `
<svg
  class="test-class svg-icon"
  data-testid="icon"
  height="24"
  width="24"
>
  <use
    xlink:href="#point"
  />
</svg>
`;

exports[`Icon задается кастомная ширина и высота 1`] = `
<svg
  class="svg-icon"
  data-testid="icon"
  height="150"
  width="100"
>
  <use
    xlink:href="#point"
  />
</svg>
`;

exports[`Icon меняется тип иконки 1`] = `
<svg
  class="svg-icon"
  data-testid="icon"
  height="24"
  width="24"
>
  <use
    xlink:href="#switch"
  />
</svg>
`;

exports[`Icon установлены дефолтная ширина и высота 1`] = `
<svg
  class="svg-icon"
  data-testid="icon"
  height="24"
  width="24"
>
  <use
    xlink:href="#point"
  />
</svg>
`;

exports[`Icon ширина и высота заданы через проп size 1`] = `
<svg
  class="svg-icon"
  data-testid="icon"
  height="32"
  width="32"
>
  <use
    xlink:href="#point"
  />
</svg>
`;
